rfd/00xx-snowflake-support.md

```diff
@@ -0,0 +1,170 @@
+      protocol: "snowflake"
+      uri: "im12345.us-east-2.aws.snowflakecomputing.com"
+```
+
+The next step would be to add teleport CA to the Snowflake user account. `tctl` could learn to export the required keys 
+in the same way as it does today for different databases (`tctl auth sign` command). The exported key must then be added 
+to Snowflake by an administrator. The procedure of adding the key to Snowflake is described [here](https://docs.snowflake.com/en/user-guide/key-pair-auth.html#step-4-assign-the-public-key-to-a-snowflake-user).
+
+### Database access support
+
+`tsh db login` should export a user private key in PCKS8 format (that's the only format supported by `snowsql` and some SDKs).
+That key should be used to sign a "Teleport JWT". Client then should send the request with generated JWT to authenticate 
+with Snowflake. Teleport can intercept HTTP communication between a client and the Snowflake. The first request send by 
+the client should be to the `/session/v1/login-request` endpoint which contains all login information as username, JWT. 
+This data can be used to authenticate user in Teleport. JWT token prepared by the client can be then read and replaces by Snowflake JWT
+generated by Teleport DB service. Teleport should use Database CA to sign JWT at that point to sign the request instead 
+of User CA to be able to authenticate with the key generated by `tctl auth sign` and imported to Snowflake. 
+In this way the real token is not accessible by a user and Teleport can still validate if the request 
+comes from a trusted source.
+
+```mermaid
+sequenceDiagram
+    participant client
+    participant Teleport
+    participant Snowflake
+    client->>Teleport: HTTP /session/v1/login-request
+    Teleport->>Teleport: RBAC check, genererate and replace JWT
+    Teleport->>Snowflake: HTTP /session/v1/login-request
+    Snowflake->>Teleport: Authentication Token
+    Teleport->>Teleport: Generate Teleport Auth token
+    Teleport->>client: Send Teleport Auth token
+```
+
+Nn each request Teleport can replace the "Teleport Auth token" with Snowflake Auth token:
+
+```mermaid
+sequenceDiagram
+    participant client
+    participant Teleport
+    participant Snowflake
+    client->>Teleport: HTTP /queries/v1/query-request
+    Teleport->>Teleport: Replace Auth header
+    Teleport->>Snowflake: HTTP /queries/v1/query-request
+    Snowflake->>Teleport: Response
+    Teleport->>client: Forward the response
+```
+
+With this approach the real session token never leaves the Teleport and any client is not able to bypass Teleport
+by extracting the session token.
+Snowflake session token needs to be refreshed every X minutes. For that purpose Teleport would have to internally refresh
+the token as client doesn't have access to the token (login response contains the time for how long the token is valid).
+
+#### Query logging
+
+Queries can be extracted from requests sent to `/queries/v1/query-request` endpoint, field `data.sqltext`.
```

Comment:
Is there protocol documentation we can link here?

Reply:
I wasn't able to find anything official 😞 Maybe Snowflake will point us to some documentation.